### PR TITLE
Implement EIP 7742 in Prague

### PR DESCRIPTION
**Pull request by petertdavies**
This PR implements EIP 7742 on top of Prague. It have done some brief tests with the [execution-spec-tests PR](https://github.com/ethereum/execution-spec-tests/pull/986) and these changes seem to work.

~This PR should not be merged since the status of EIP 7742 is very uncertain.~

Please check that EIP 7742 is actually going into `prague-devnet-5` before merging.

#### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Comment by spencer-tb** [comment]
Thanks again for adding this :D

We were wondering the following should be added too:
https://github.com/spencer-tb/execution-specs/commit/07f4774873cf25e062d925f43c740f6fdf8ada24#diff-1ead0e19bcf5b72f6e0ff1f07a9d44a25a738020803d5ea42839da1e3d58a8c5R433-R434

Checking for contract creating blob txs!

File: src/ethereum_spec_tools/evm_tools/t8n/__init__.py [hunk 1]
```diff
@@ -357,7 +357,9 @@ def apply_body(self) -> None:
                     env, tx
                 )
 
-                if self.fork.is_after_fork("ethereum.cancun"):
+                if self.fork.is_after_fork(
+                    "ethereum.cancun"
+                ) and not self.fork.is_after_fork("ethereum.prague"):
                     blob_gas_used += self.fork.calculate_total_blob_gas(tx)
```
Comment (gurukamath) [inline src/ethereum_spec_tools/evm_tools/t8n/__init__.py:363]: Incrementing the `blob_gas_used` has to happen post Prague as well I believe. It's only the `MAX_BLOB_GAS_PER_BLOCK` check that is removed

File: src/ethereum/prague/vm/gas.py [hunk 2]
```diff
@@ -286,10 +288,11 @@ def calculate_excess_blob_gas(parent_header: Header) -> U64:
     parent_blob_gas = (
         parent_header.excess_blob_gas + parent_header.blob_gas_used
     )
-    if parent_blob_gas < TARGET_BLOB_GAS_PER_BLOCK:
+    blob_gas = GAS_PER_BLOB * header.target_blobs_per_block
```
Comment (gurukamath) [inline src/ethereum/prague/vm/gas.py:291]: Does naming this variable `blob_gas_target` make sense?

File: src/ethereum_spec_tools/evm_tools/t8n/env.py [hunk 3]
```diff
@@ -114,12 +116,25 @@ def read_excess_blob_gas(self, data: Any, t8n: "T8N") -> None:
             self.parent_excess_blob_gas + self.parent_blob_gas_used
         )
 
-        target_blob_gas_per_block = t8n.fork.TARGET_BLOB_GAS_PER_BLOCK
+        if "currentTargetBlobsPerBlock" in data:
```
Comment (gurukamath) [inline src/ethereum_spec_tools/evm_tools/t8n/env.py:119]: For post-Prague forks, `currentTargetBlobsPerBlock` seems to be a mandatory field. So, if that is not available in the data, I believe the tools should throw an exception

**Comment by marioevz** [comment]
Created https://github.com/petertdavies/execution-specs/pull/4 to implement changes in https://github.com/ethereum/EIPs/pull/9122.

On a side note, I'm seeing zero value in the result's field `blobGasUsed`, regardless of whether a blob is included in the transactions or not, I'm not sure if this is derived from the changes in this PR though @gurukamath

**Comment by gurukamath** [comment]
> On a side note, I'm seeing zero value in the result's field `blobGasUsed`, regardless of whether a blob is included in the transactions or not, I'm not sure if this is derived from the changes in this PR though @gurukamath

Could [this](https://github.com/ethereum/execution-specs/pull/1045#discussion_r1876499398) be the issue?

File: src/ethereum_spec_tools/evm_tools/t8n/__init__.py (hunk 1 above)
Comment (petertdavies) [reply src/ethereum_spec_tools/evm_tools/t8n/__init__.py:361]: Fixed

File: src/ethereum_spec_tools/evm_tools/t8n/env.py (hunk 3 above)
Comment (petertdavies) [reply src/ethereum_spec_tools/evm_tools/t8n/env.py]: As currently written, we don't take into account what the current fork is at this point. This is an issue for the t8n refactor/rewrite to deal with.

**Comment by gurukamath** [comment]
LGTM!

**Pull request merged**